etc/afpd/afpstats.c

```diff
@@ -56,34 +99,75 @@ static gpointer afpstats_thread(gpointer _data _U_)
-
-    if (!(bus = dbus_g_bus_get_private(DBUS_BUS_SYSTEM, ctxt, &error))) {
+    if (!(bus = g_bus_get_sync(G_BUS_TYPE_SYSTEM,
+                               NULL,
+                               &error))) {
         LOG(log_error, logtype_afpd,"Couldn't connect to system bus: %s", error->message);
         return NULL;
     }
 
-    if (!(bus_proxy = dbus_g_proxy_new_for_name(bus, "org.freedesktop.DBus",
-                                                "/org/freedesktop/DBus",
-                                                "org.freedesktop.DBus"))) {
+#if 0
+    if (!(bus_proxy = g_dbus_proxy_new_sync(bus,
+                                            G_DBUS_PROXY_FLAGS_NONE,
+                                            NULL, /* GDBusInterfaceInfo */
+                                            "org.freedesktop.DBus",
+                                            "/org/freedesktop/DBus",
+                                            "org.freedesktop.DBus",
+                                            NULL, /* GCancellable */
+                                            &error))) {
         LOG(log_error, logtype_afpd,"Couldn't create bus proxy");
         return NULL;
     }
 
-    if (!dbus_g_proxy_call(bus_proxy, "RequestName", &error,
-                           G_TYPE_STRING, "org.netatalk.AFPStats",
-                           G_TYPE_UINT, 0,
-                           G_TYPE_INVALID,
-                           G_TYPE_UINT, &request_name_result,
-                           G_TYPE_INVALID)) {
+    if (!g_dbus_proxy_call_sync(bus_proxy,
+                                "RequestName",
+                                g_variant_new("(as)", "org.netatalk.AFPStats", 0),
+                                G_DBUS_CALL_FLAGS_NONE,
+                                -1, /* timeout */
+                                NULL,
+                                &error)) {
         LOG(log_error, logtype_afpd, "Failed to acquire DBUS name: %s", error->message);
         return NULL;
     }
+#endif
+
+    static const gchar introspection_xml[] =
+        "<node>"
+        "  <interface name='org.netatalk.AFPStats'>"
+        "    <method name='GetUsers'>"
+        "      <arg name='ret' type='as' direction='out'/>"
+        "    </method>"
+        "  </interface>"
+        "</node>";
+
+    introspection_data = g_dbus_node_info_new_for_xml (introspection_xml, NULL);
+    g_assert (introspection_data != NULL);
 
     AFPStatsObj *obj = g_object_new(AFPSTATS_TYPE_OBJECT, NULL);
-    dbus_g_connection_register_g_object(bus, "/org/netatalk/AFPStats", G_OBJECT(obj));
+
+    registration_id = g_dbus_connection_register_object(bus,
+                                      "/org/netatalk/AFPStats",
+                                      introspection_data->interfaces[0],
+                                      &interface_vtable,
+                                      G_OBJECT(obj),
+                                      NULL,
```

Comment:
It's a good idea to explicitly keep the object "alive" for as long as it's registered:

```suggestion
                                      g_object_ref (obj),
                                      g_object_unref,
```